src/modules/app/browser-init.service.ts

```diff
@@ -144,15 +144,20 @@ export class BrowserInitService extends InitService {
    * @private
    */
   private async loadAppState(): Promise<boolean> {
-    const state = this.transferState.get<any>(InitService.NGRX_STATE, null);
-    this.transferState.remove(InitService.NGRX_STATE);
-    this.store.dispatch(new StoreAction(StoreActionTypes.REHYDRATE, state));
-    return lastValueFrom(
-      this.store.select(coreSelector).pipe(
-        find((core: any) => isNotEmpty(core)),
-        map(() => true),
-      ),
-    );
+    // The app state can be transferred only when SSR and CSR are using the same base url for the REST API
```

Comment:
A separate REST URL could be a performance benefit during SSR, but if we discard all cached requests we'll _always_ do double work and CSR performance will suffer.

Could we consider just running a search/replace on the transfer state to replace the SSR URL with the CSR URL? 
- This shouldn't be too heavy of an operation so I think the net impact on performance could still be positive
- I did a very quick check and replacing all URLs right after the full HTML has been generated takes at most 0.25% of the total time needed for SSR
    - Only tested for `/home`, not authenticated
    - SSR took ~400ms, HTML search replace took between 0 and 1ms

Otherwise we should at least exclude the transfer state entirely; right now it's sent but never read.

Reply:
I understand your suggestion, but I believe implementing a search/replace operation for URLs might not be a good idea due to performance concerns. Even though the initial tests show a minimal impact, this might not be the case for all routes or under different conditions and it could become a bottleneck.
Moreover adding an extra step in the Server-side rendering process can increase the overall latency, negatively affecting the user experience, particularly for pages with larger amount of request.
Honestly I don't see a real issue to fetch again the request during the CSR.

I'll try to exclude the transfer state entirely

Reply:
> Honestly I don't see a real issue to fetch again the request during the CSR

My main concern is that it has a significant impact on perceived performance, especially on slower connections.
This counteracts our push towards decreasing the amount of REST requests we send ~ https://github.com/DSpace/dspace-angular/issues/3110#issuecomment-2207295962.

I've made a quick branch where you can toggle both things: [ybnd/dspace-angular → task/main/DURACOM-288](https://github.com/ybnd/dspace-angular/commits/task/main/DURACOM-288)
- `ui.transferState` → whether NGRX state is transferred from SSR to CSR
- `replaceRestUrl` → whether the SSR URL is replaced after SSR is done

As far as I can tell, a simple RegEx to replace SSR URLs scales a lot better than DSpace itself
- A `/search` page with 1000 results (!) takes ~50000ms to build with SSR on my machine
- Replacing all URLs (~16000) takes ~20ms

On the other hand, for a `/search` page with 100 results, Lighthouse reports about double the blocking time if we don't transfer state on CSR.

<details>
  <summary>With URL replacing & state transfer</summary>
  <img src="https://github.com/user-attachments/assets/ff410fdb-e776-4e38-a6ba-6bdde80a6293">
</details>
<details>
  <summary>Without state transfer</summary>
  <img src="https://github.com/user-attachments/assets/bf548bf8-a517-44a1-9130-31eb00645f5d">
</details>